### PR TITLE
hoppscotch: init at 23.12.1

### LINKED ISSUE #267059: Package request: Hoppscotch Desktop App
Open source API development ecosystem. Essentially an open-source alternative to Postman.

**Metadata**

* homepage URL: https://hoppscotch.com/download
* source URL: https://github.com/hoppscotch/hoppscotch
* license: mit
* platforms: linux, darwin

### PR DESCRIPTION
**Pull request by GetPsyched**
## Description of changes

Similar to Postman in terms of core functionality.

Copied from upstream: Hoppscotch is a lightweight, web-based API development suite. It was built the ground up with ease of use and accessibility in mind providing the functionality needed for API developers with minimalist, UI.

Closes #267059

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by GetPsyched** [comment]
Marked as a draft since upstream has a bug that needs to be resolved.

Note: I've put the package in `bin.nix` instead of `default.nix` to make way for this to be packaged from source later down the line. It's still called `hoppscotch` and not `hoppscotch-bin` in `all-packages.nix` since the source version doesn't exist yet; once it does, both alternatives should exist.

If there's anything wrong/unconventional with this approach, I'm fine with changing my PR structure.

**Comment by MikaelFangel** [comment]
Fixes: #267059

File: pkgs/development/web/hoppscotch/bin.nix [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+      all the functionality needed for API developers with minimalist,
+      unobtrusive UI.
+    '';
+    homepage = "https://hoppscotch.com";
+    downloadPage = "https://hoppscotch.com/downloads";
+    changelog = "https://hoppscotch.com/changelog";
+    license = licenses.mit;
+    maintainers = with maintainers; [ getpsyched ];
+    mainProgram = pname;
+    platforms = platforms.all;
```
Comment (MikaelFangel) [inline pkgs/development/web/hoppscotch/bin.nix:58]: Isn't the `meta.platforms` only Linux as the derivation downloads a Linux AppImage?

**Comment by GetPsyched** [comment]
> Fixes: https://github.com/NixOS/nixpkgs/issues/267059

Huh, interesting. I searched for open issues regarding this but didn't find any. Guess GH search was borked for a moment.

File: pkgs/development/web/hoppscotch/bin.nix (hunk 1 above)
Comment (GetPsyched) [reply pkgs/development/web/hoppscotch/bin.nix:58]: Oops, my bad. I incorrectly assumed the AppImage to be platform agnostic. Pushing a fix real quick.

Comment (GetPsyched) [reply pkgs/development/web/hoppscotch/bin.nix]: I think it might still be incorrect. Since the AppImage specifies the architecture too?

**Comment by GetPsyched** [comment]
EDIT: Login is borked. I'm testing this (unsure if it's a derivation or upstream issue)

**Comment by nephalemsec** [comment]
Builds fine for me, but the application launches as a blank screen. Assuming it is something to do with video drivers + wayland/hyprland.

I've attempted building from .deb instead of .appimage and builds and looks fine, but can't send any requests :exploding_head: 

```nix
{ stdenv, dpkg, curl, glib, glibc, libudev0-shim, nghttp2, libsoup, cairo, pango, gtk3-x11, gdk-pixbuf, webkitgtk, openssl, desktop-file-utils, hicolor-icon-theme
, autoPatchelfHook, fetchurl, lib, makeDesktopItem }:
# Dependencies
# https://aur.archlinux.org/packages/hoppscotch-app-bin
let
  name = "hoppscotch";
  version = "23.12.2";

  desktopItem = makeDesktopItem {
    categories = [ "Development" ];
    desktopName = "Hoppscotch";
    exec =
      "WEBKIT_DISABLE_COMPOSITING_MODE=1 hoppscotch-app %U"; # https://github.com/tauri-apps/tauri/issues/4315
    icon = name;
    inherit name;
  };

  icon = fetchurl {
    url =
      "https://raw.githubusercontent.com/hoppscotch/hoppscotch/20${version}/packages/hoppscotch-common/public/logo.svg";
    hash = "sha256-Njbc+RTKSOziXo0H2Mv7RyNI5CLZNkJLUr/PatyrK9E=";
  };
in stdenv.mkDerivation rec {
  inherit name version;

  src = fetchurl {
    url =
      #"https://github.com/hoppscotch/releases/releases/download/v${version}-1/Hoppscotch_linux_x64.deb";
      "https://github.com/liudonghua123/hoppscotch-app/releases/latest/download/hoppscotch-app-linux-${version}_amd64.deb";
    hash = "sha256-Ajp2Oiox3v1OTIZg/vtZj/eftTUHieEvwJfpXxoU2jI=";
  };

  dontBuild = true;
  dontConfigure = true;
  dontWrapGApps = true;

  nativeBuildInputs = [ dpkg autoPatchelfHook ];
  buildInputs = [ glib libsoup cairo pango gtk3-x11 gdk-pixbuf webkitgtk openssl desktop-file-utils hicolor-icon-theme  ];

  unpackPhase = "dpkg-deb -x $src $out";

  installPhase = ''
    runHook preInstall
    mkdir -p $out/bin $out/share
    cp -av $out/usr/bin/hoppscotch-app $out/bin/hoppscotch-app
    cp -av $out/usr/share/* $out/share
    rm -rf $out/usr
    runHook postInstall
  '';

  extraInstallCommands = ''
    install -D ${icon} $out/share/icons/hicolor/scalable/apps/${name}.svg

    mkdir -p $out/share/applications
    cp -r ${desktopItem} $out/share/applications/${name}.desktop
  '';

  meta = with lib; {
    description = "👽 Open-source API development ecosystem";
    longDescription = ''
      Hoppscotch is a lightweight, web-based API development suite.
    '';
    homepage = "https://hoppscotch.com";
    downloadPage = "https://hoppscotch.com/downloads";
    changelog = "https://hoppscotch.com/changelog";
    license = licenses.mit;
    platforms = platforms.linux;
  };
}
```

**Comment by GetPsyched** [comment]
Thanks for sharing your WIP attempt at the .deb option.

AFAICT, these issues are likely from upstream since the desktop app is still not stable. Sadly, the only option we have is to wait.

**Comment by GetPsyched** [comment]
Well, then.

**Pull request closed without merging**